### PR TITLE
Support Rails 6.1

### PR DESCRIPTION
**Pull request by JonRowe**
This branch will become the root of Rails 6.1 development. The plan would be to launch this as a new version when Rails 6.1 is released, likely a major version as Rails versions supported will change.

*Edit* This will not be merged until we are ready to release the next version, it will be used as an available branch for those working on 6.1.

Todo:
- [x] View specs have changed behaviour surrounding helper modules, investigate and fix.
- [x] A controller spec has a broken config somehow
- [x] Generated controller spec "uploads" fails

### CODE REVIEW

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 1]
```diff
@@ -66,6 +68,7 @@ module ::ApplicationHelper; end
           group = RSpec::Core::ExampleGroup.describe('bars/new.html.erb') do
             def self.helper(*); end # Stub method
           end
+        allow(group).to receive(:helper)
```
Comment (pirj) [inline spec/rspec/rails/example/view_example_group_spec.rb:71]: The indentation is off.

**Review by pirj — COMMENTED** [review]
Looks good.
I've missed the last build results, wondering how many things are broken.

**Comment by benoittgt** [comment]
I do not understand why it is not able to find 3.10.0.pre. I have no issue locally to find them.

```
Using rspec 3.10.0.pre from source at `/Users/bti/code/rspec-dev/repos/rspec`

**Comment by JonRowe** [comment]
@benoittgt @pirj My intent for this branch is to use it as dev / maintenance root branch so could you PR to it rather than push commits directly, especially for WIP commits...

**Comment by JonRowe** [comment]
@benoittgt I'm going to unpick your commits to a PR, from now on can everyone PR to this branch and that should be the last force push

**Comment by benoittgt** [comment]
I just lauch a CI build for Rails v6.1.0.rc2 https://github.com/rspec/rspec-rails/actions/runs/396008092

**Comment by benoittgt** [comment]
Ok. 3 commits need to be cherry pick if we want to make it workable https://github.com/rspec/rspec-rails/compare/rails-6-1-dev...rails-6.1.0.rc2-dev

1. https://github.com/rspec/rspec-rails/pull/2400/commits/05f3a2caa3244b0ebcbcd3e07227ee014215f7e6
2. https://github.com/rspec/rspec-rails/commit/7300e06a07e31391e7de5519ac258402a5c8e914
3. https://github.com/rspec/rspec-rails/commit/5737b93361f24bf0994dbf94ea8e8562a7c8fbcb

We will still have the current issue in the CI with Rails master and dependencies resolving. Should we have rails-6.1.0.rc2 + rails master in the CI for this branch?

**Comment by pirj** [comment]
> Should we have rails-6.1.0.rc2 + rails master in the CI for this branch?

Sure, why not both? As long if we don't set `required` for those that consistently fail.

**Comment by benoittgt** [comment]
> Sure, why not both? As long if we don't set required for those that consistently fail.

Yes I am in favor of both.

**Comment by benoittgt** [comment]
Rails 6.1 has been released. I just updated to CI and I think we are good enough to release. https://github.com/rspec/rspec-rails/pull/2408

I will look at https://github.com/rspec/rspec-rails/issues/2410

**Comment by benoittgt** [comment]
CI is green. I think we should merge https://github.com/rspec/rspec-rails/pull/2408 and release.

**Comment by pirj** [comment]
@benoittgt I see failures against `master`.
Does it make sense to add Rails 6.1 to the matrix?

**Comment by benoittgt** [comment]
But 6.1 is in the matrix https://github.com/rspec/rspec-rails/pull/2408/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR38 ? 🤔😊

**Comment by pirj** [comment]
Ah, cool! I didn't see that other PR.

**Comment by JonRowe** [comment]
I'm going to merge this with a view to releasing today or more likely over the weekend.

There is a problem with active mail box on Ruby 3.0.0-preview2 but thats a preview so I'm happy to mark it as "allow failure" and try to fix it in a patch release for Ruby 3.0.0 final later on.

File: Changelog.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+### Rails 6.1 Development
+[Full Changelog](https://github.com/rspec/rspec-rails/compare/main...rails-6-1-dev)
```
Comment (pirj) [inline Changelog.md:2]: `v4.0.1...4.1.0`?

**Review by pirj — APPROVED** [review]
:+1: 

Will #2412 and a fix for #2417 (if it's a real issue) be part of the release?

**Comment by JonRowe** [comment]
I'm holding off merging until later today, because I'm going to sort out the versions when I do it (ready for release)

File: Changelog.md (hunk 2 above)
Comment (JonRowe) [reply Changelog.md:2]: See below

**Comment by rafaelfranca** [comment]
https://github.com/rails/rails/pull/40780 was merged and will be included in 6.1.1, so I think you don't need to define the missing method anymore.

**Comment by benoittgt** [comment]
> There is a problem with active mail box on Ruby 3.0.0-preview2 but thats a preview so I'm happy to mark it as "allow failure" and try to fix it in a patch release for Ruby 3.0.0 final later on.

I dig into tonight. It is related to https://github.com/rails/rails/issues/40756 and https://github.com/ruby/ruby/pull/3861. I will follow both issues. But as you say it can be ignored at the moment.

**Comment by benoittgt** [comment]
Thanks @rafaelfranca . I will probably close https://github.com/rspec/rspec-rails/pull/2412 then.

**Comment by pirj** [comment]
@benoittgt, what are your thoughts on #2410? Should we patch broken Rails `6.1.0-rc2`..`6.10` as in https://github.com/rspec/rspec-rails/pull/2412?
If we declare them as unsupported in our gemspec, won't Bundler resolve the "matching" `rspec-rails` version to 4.0.x that won't work with Rails 6.1, and induce user confusion?

**Comment by benoittgt** [comment]
@pirj I was curious about this. My patch is not very good and I find it weird to include it by default in a very small release of Rails. Like include `ActiveSupport::Testing::Assertions` only between Rails 6.0 included and 6.1. I prefer people to update Rails to 6.1.1 + rspec-rails 4, no?

**Comment by pirj** [comment]
@benoittgt Agree. Maybe we can catch `NoMethodError` for Rails 6.1.0, check if it matches `assert_nothing_raised` and fail with a detailed message that users should update to 6.1.1?

**Comment by benoittgt** [comment]
> @benoittgt Agree. Maybe we can catch NoMethodError for Rails 6.1.0, check if it matches assert_nothing_raised and fail with a detailed message that users should update to 6.1.1?

Isn't something we do? I would like to know if @JonRowe would be ok with this. I find this a little bit hacky but if it can reduce support and dead end for user that try to upgrade. 🤷

**Comment by erikaxel** [comment]
Hi, just want to mention that as a user of the system I would find it very helpful to have this hack for 6.1, especially since 6.1.1 is not released yet. Last time it took approximately 2 months for the Rails .1 to come out, and it would be great to have an official gem version of rspec that works with 6.1 meanwhile. Also thinking about new users trying out Rails and then suddenly get stuck if they would like to try out rspec instead of mini-test. PS: Thanks for the great work!

**Comment by pirj** [comment]
Just for the record https://github.com/rails/rails/milestone/76

**Comment by JonRowe** [comment]
Are positive this isn't an include order thing? After all including the module fixes it... People may still be using the mini test assertions with our runner.

**Comment by benoittgt** [comment]
I reran the CI this aftertoon to test 6.1.1 (released yesterday) and we do not have new issues with this branch.

**Comment by pirj** [comment]
Just a few tickets left! https://github.com/rspec/rspec-rails/milestone/20 🙌

**Review by pirj — APPROVED** [review]
What is left is:
 - https://github.com/rspec/rspec-rails/issues/2426 (no clear reproduction case)
 - https://github.com/rspec/rspec-rails/issues/2410 (I doubt we need to provide access to `assert_nothing_raised` out of the box given we provide `not_to raise_error`)

@JonRowe @benoittgt Do you think there are any roadblocks left to releasing 4.1?

**Comment by benoittgt** [comment]
I see no road block for a release.

I was thinking yesterday about starting the blog post for the release.

**Comment by JonRowe** [comment]
I'm worried that #2410 is a symptom of a module missing we should include

**Comment by pirj** [comment]
Is https://github.com/rspec/rspec-rails/issues/2451 a release blocker?

Is there anything else left?

**Comment by benoittgt** [comment]
I think #2451 could be added in the next minor or patch release.

**Comment by doits** [comment]
:+1: for a Rails 6.1 compatible release!

**Comment by pirj** [comment]
I believe we've even achieved our stretch goal by fixing `name`/`method_name`.
Any roadblocks to releasing it?

**Comment by JonRowe** [comment]
No, just need to find some time to sort out the versioning, as this will be a new major release theres some cleanup work to do around ensuring everything eligible from `main` is synced across to `4-0-maintenance` and any patch releases sent out, then this is merged and we start `5-0-maintenance` and release from there etc

**Comment by JonRowe** [comment]
We need to add a deprecation warning to this branch so that people using it know to switch to `main` or a release coming up this weekend.

**Comment by hasghari** [comment]
When my project pointed to the `rails-6-1-dev` branch, everything worked as expected but after updating `rspec-rails` to the released `4.1.0` version, I get the following error message:
```
Failure/Error: let(:image) { fixture_file_upload('/images/slider-buttons.psd') }

NoMethodError:
  undefined method `file_fixture_path' for RSpec::Rails::FixtureFileUploadSupport::RailsFixtureFileWrapper:Class
  Did you mean?  fixture_path
```

**Comment by gueorgui** [comment]
@hasghari I had the same issue a while ago — though I think it's related to the Rails upgrade rather than to the rspec-rails update. The way to resolve it is to add this to your spec:

```ruby
include ActionDispatch::TestProcess::FixtureFile
```

(According to [this SO answer](https://stackoverflow.com/a/66134337/12288625))

**Comment by JonRowe** [comment]
> When my project pointed to the rails-6-1-dev branch, everything worked as expected but after updating rspec-rails to the released 4.1.0 version, I get the following error message:
4.1.0 does not contain this change, 5.0.0 will.

**Pull request merged**